File: rd-net/RdFramework/Impl/Protocol.cs
using System;
using System.Collections.Generic;
using System.Threading;
using JetBrains.Annotations;
using JetBrains.Collections.Synchronized;
using JetBrains.Collections.Viewable;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.Rd.Base;

namespace JetBrains.Rd.Impl
{
  public class Protocol : IProtocol
  {
    public static readonly ILog Logger = Log.GetLog("protocol");
    public static readonly ILog InitLogger = Logger.GetSublogger("INIT");
    public static LogWithLevel? InitTrace = InitLogger.WhenTrace();

    /// <summary>
    /// Should match textual RdId of protocol intern root in Kotlin/js/cpp counterpart
    /// </summary>
    const string ProtocolInternRootRdId = "ProtocolInternRoot";
    const string ContextHandlerRdId = "ProtocolContextHandler";
    internal const string ProtocolExtCreatedRdId = "ProtocolExtCreated";
    
    /// <summary>
    /// Should match whatever is in rd-gen for ProtocolInternScope
    /// </summary>
    const string ProtocolInternScopeStringId = "Protocol";

    public Lifetime Lifetime { get; }

    public RdEntitiesRegistrar RdEntitiesRegistrar { get; }
    
    private readonly Protocol? myParentProtocol;
    private readonly Dictionary<string, object> myExtensions = new();

    public Protocol(string name, ISerializers serializers, IIdentities identities, IScheduler scheduler, 
      IWire wire, Lifetime lifetime, params RdContextBase[] initialContexts) 
      : this(name, serializers, identities, scheduler, wire, lifetime, null, null, initialContexts)
    { }

    internal Protocol(string name, ISerializers serializers, IIdentities identities, IScheduler scheduler,
      IWire wire, Lifetime lifetime, Protocol? parentProtocol, RdSignal<ExtCreationInfo>? parentExtConfirmation = null, params RdContextBase[] initialContexts)
    {
      Lifetime = lifetime;
      Name = name ?? throw new ArgumentNullException(nameof(name));
      Location = new RName(name);

      Serializers = serializers ?? throw new ArgumentNullException(nameof(serializers));
      Identities = identities ?? throw new ArgumentNullException(nameof(identities));
      Scheduler = scheduler ?? throw new ArgumentNullException(nameof(scheduler));
      Wire = wire ?? throw new ArgumentNullException(nameof(wire));
      myParentProtocol = parentProtocol;
      RdEntitiesRegistrar = parentProtocol?.RdEntitiesRegistrar ?? new RdEntitiesRegistrar();
      SerializationContext = parentProtocol?.SerializationContext ?? new SerializationCtx(this, new Dictionary<string, IInternRoot<object>>() {{ProtocolInternScopeStringId, CreateProtocolInternRoot(lifetime)}});
      Contexts = parentProtocol?.Contexts ?? new ProtocolContexts(SerializationContext);
      wire.Contexts = Contexts;
      if (parentProtocol?.SerializationContext == null)
        SerializationContext.InternRoots[ProtocolInternScopeStringId].BindTopLevel(lifetime, this, ProtocolInternRootRdId);
      foreach (var rdContextBase in initialContexts) rdContextBase.RegisterOn(Contexts);
      if (parentProtocol?.Contexts == null)
        BindContexts(lifetime);
      OutOfSyncModels = new ViewableSet<RdExtBase>();
      ExtCreated = parentProtocol?.ExtCreated ?? new Signal<ExtCreationInfoEx>();
      ExtConfirmation = parentExtConfirmation ?? this.CreateExtSignal(identities);
      ExtIsLocal = new ThreadLocal<bool>(() => false);
      ExtConfirmation.Advise(lifetime, message =>
      {
        ExtCreated.Fire(new ExtCreationInfoEx(message, ExtIsLocal.Value));
      });
      using (AllowBindCookie.Create()) 
        ExtConfirmation.BindTopLevel(lifetime, this, ProtocolExtCreatedRdId);

      if (wire is IWireWithDelayedDelivery wireWithMessageBroker)
        wireWithMessageBroker.StartDeliveringMessages();
    }

    public bool TryGetSerializationContext(out SerializationCtx ctx)
    {
      ctx = SerializationContext;
      return true;
    }

    private InternRoot<object> CreateProtocolInternRoot(Lifetime lifetime)
    {
      var root = new InternRoot<object>();
      root.RdId = Identities.Mix(RdId.Nil, ProtocolInternRootRdId);
      
      return root;
    }

    private void BindContexts(Lifetime lifetime)
    {
      Contexts.RdId = Identities.Mix(RdId.Nil, ContextHandlerRdId);
      using (AllowBindCookie.Create()) 
      {
        Contexts.PreBind(lifetime, this, ContextHandlerRdId);
        Contexts.Bind();
      }
    }
    
    internal void SubmitExtCreated(ExtCreationInfo info)
    {
      Assertion.Assert(!ExtIsLocal.Value, "!ExtIsLocal");
      ExtIsLocal.Value = true;
      try
      {
        ExtConfirmation.Fire(info);
      }
      finally
      {
        ExtIsLocal.Value = false;
      }
    }
      
    public string Name { get; }
    
    public IWire Wire { get; }
    public ISerializers Serializers { get; }
    public IIdentities Identities { get; }
    public IScheduler Scheduler { get; }
    public SerializationCtx SerializationContext { get; }
    public ViewableSet<RdExtBase> OutOfSyncModels { get; }

    public ProtocolContexts Contexts { get; }
    
    public ISignal<ExtCreationInfoEx> ExtCreated { get; }
    
    private RdSignal<ExtCreationInfo> ExtConfirmation { get; }

    private ThreadLocal<bool> ExtIsLocal { get; }

    [PublicAPI] public bool ThrowErrorOnOutOfSyncModels = true;
    
    
    public RName Location { get; }
    IProtocol IRdDynamic.TryGetProto() => this;

    public virtual T? GetExtension<T>() where T : RdExtBase
    {
      var parentProtocol = myParentProtocol;
      if (parentProtocol != null)
        return parentProtocol.GetExtension<T>();
      
      lock (myExtensions)
      {
        return myExtensions.TryGetValue(typeof(T).Name, out var extension) ? (T)extension : default;
      }
    }
    
    public virtual T GetOrCreateExtension<T>(Func<T> create) where T : RdExtBase
    {
      if (create == null) throw new ArgumentNullException(nameof(create));
      
      var parentProtocol = myParentProtocol;
      if (parentProtocol != null)
        return parentProtocol.GetOrCreateExtension(create);

      lock (myExtensions)
      {
        var name = typeof(T).Name;
        if (myExtensions.TryGetValue(name, out var existing))
        {
          var val = existing.NotNull("Found null value for key: '{0}'", name) as T;
          Assertion.Require(val != null, $"Found bad value for key '{name}'. Expected type: '{typeof(T).FullName}', actual:'{existing.GetType().FullName}");
          return val;
        }

        var res = create().NotNull("'Create' result must not be null");

        myExtensions[name] = res;
        if (res is IRdBindable rdBindable)
        {
          rdBindable.Identify(Identities, Identities.Mix(RdId.Root, name));
          rdBindable.PreBind(Lifetime, this, name);
          rdBindable.Bind();
        }

        return res;
      }
    }
  }
}

File: rd-kt/rd-framework/src/main/kotlin/com/jetbrains/rd/framework/Protocol.kt
package com.jetbrains.rd.framework

import com.jetbrains.rd.framework.base.AllowBindingCookie
import com.jetbrains.rd.framework.base.RdExtBase
import com.jetbrains.rd.framework.base.bindTopLevel
import com.jetbrains.rd.framework.impl.InternRoot
import com.jetbrains.rd.framework.impl.ProtocolContexts
import com.jetbrains.rd.framework.impl.RdSignal
import com.jetbrains.rd.util.Sync
import com.jetbrains.rd.util.collections.SynchronizedSet
import com.jetbrains.rd.util.getLogger
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.reactive.*
import com.jetbrains.rd.util.string.RName
import java.util.concurrent.ConcurrentHashMap
import kotlin.reflect.KClass


class Protocol internal constructor(
    override val name: String,
    override val serializers: ISerializers,
    override val identity: IIdentities,
    override val scheduler: IScheduler,
    override val wire: IWire, //to initialize field with circular dependencies
    override val lifetime: Lifetime,
    private val parentProtocol: Protocol?,
    parentExtConfirmation: RdSignal<ExtCreationInfo>? = null,
    vararg initialContexts: RdContext<*>
) : IRdDynamic, IProtocol {

    @Deprecated("Backward compatible implementation for AWS plugin compile against Rider SDK 2019.2")
    constructor(serializers: ISerializers,
                identity: IIdentities,
                scheduler: IScheduler,
                wire: IWire,
                lifetime: Lifetime) : this("Noname-Please-Specify-Name", serializers, identity, scheduler, wire, lifetime)

    constructor(name: String,
                serializers: ISerializers,
                identity: IIdentities,
                scheduler: IScheduler,
                wire: IWire, //to initialize field with circular dependencies
                lifetime: Lifetime,
                vararg initialContexts: RdContext<*>) : this(name, serializers, identity, scheduler, wire, lifetime, null, null, *initialContexts)

    override val location: RName = RName(name)
    override val outOfSyncModels: ViewableSet<RdExtBase> = ViewableSet(SynchronizedSet())

    override val isMaster: Boolean = identity.dynamicKind == IdKind.Client

    val rdEntitiesRegistrar: RdEntitiesRegistrar = parentProtocol?.rdEntitiesRegistrar ?: RdEntitiesRegistrar()

    companion object {
        val logCategory = "protocol"
        fun sublogger(subcategory: String) = getLogger("$logCategory.$subcategory")
        val initializationLogger = sublogger("INIT")
    }

    override val protocol: IProtocol get() = this
    override val serializationContext: SerializationCtx = parentProtocol?.serializationContext ?: SerializationCtx(serializers, identity, mapOf("Protocol" to InternRoot<Any>().also {
        it.rdid = identity.mix(RdId.Null, "ProtocolInternRoot")
    }))

    override val contexts: ProtocolContexts = parentProtocol?.contexts ?: ProtocolContexts(serializationContext)

    override val extCreated: ISignal<ExtCreationInfoEx>

    private val extConfirmation: RdSignal<ExtCreationInfo>
    private val extIsLocal: ThreadLocal<Boolean>

    private val extensionPerClassLocks = ConcurrentHashMap<KClass<*>, Any>()
    private val extensions = ConcurrentHashMap<KClass<*>, Any>()

    init {
        wire.setupContexts(contexts)

        if(parentProtocol?.serializationContext == null) {
            serializationContext.internRoots.getValue("Protocol").bindTopLevel(lifetime, this, "ProtocolInternRoot")
        }

        initialContexts.forEach {
            contexts.registerContext(it)
        }

        if (parentProtocol?.contexts == null) {
            contexts.also {
                it.rdid = identity.mix(RdId.Null, "ProtocolContextHandler")
                AllowBindingCookie.allowBind {
                    it.bindTopLevel(lifetime, this, "ProtocolContextHandler")
                }
            }
        }

        extCreated = parentProtocol?.extCreated ?: Signal()
        extConfirmation = parentExtConfirmation ?: createExtSignal(identity).also { signal ->
            val protocolScheduler = scheduler
            signal.scheduler = (protocolScheduler as? ISchedulerWithBackground)?.backgroundScheduler ?: protocolScheduler
        }
        extIsLocal = ThreadLocal.withInitial { false }
        AllowBindingCookie.allowBind {
            extConfirmation.bindTopLevel(lifetime, this, "ProtocolExtCreated")
            extConfirmation.advise(lifetime) { info ->
                // triggered both from local and remote sides
                extCreated.fire(ExtCreationInfoEx(info, extIsLocal.get()))
            }
        }

        if (wire is IWireWithDelayedDelivery)
            wire.startDeliveringMessages()
    }

    internal fun submitExtCreated(info: ExtCreationInfo) {
        require(extIsLocal.get() == false) { "!extIsLocal" }
        extIsLocal.set(true)
        try {
            extConfirmation.fire(info)
        } finally {
            extIsLocal.set(false)
        }
    }

    override fun <T: RdExtBase> getOrCreateExtension(clazz: KClass<T>, create: () -> T): T {
        parentProtocol?.let {
            return it.getOrCreateExtension(clazz, create)
        }

        val lock = extensionPerClassLocks.getOrPut(clazz) { Any() }
        Sync.lock(lock) {
            val res = extensions[clazz] ?: run {
                val newExtension = create()
                extensions[clazz] = newExtension
                val declName = clazz.simpleName ?: error("Can't get simple name for class $clazz")
                newExtension.identify(identity, identity.mix(RdId.Null, declName))
                newExtension.bindTopLevel(lifetime, this, declName)
                newExtension
            }
            return castExtension(res, clazz)
        }
    }

    override fun <T : RdExtBase> tryGetExtension(clazz: KClass<T>): T? {
        parentProtocol?.let {
            return it.tryGetExtension(clazz)
        }

        val lock = extensionPerClassLocks.getOrPut(clazz) { Any() }
        Sync.lock(lock) {
            val res = extensions[clazz] ?: return null
            return castExtension(res, clazz)
        }
    }
    
    private fun <T: Any> castExtension(value: Any, clazz: KClass<T>): T {
        @Suppress("UNCHECKED_CAST")
        return value as? T
            ?: error("Wrong class found in top level extension, expected `${clazz.simpleName}` but found `${value::class.simpleName}`")  
    } 
}
